AWSAppSyncClient/AWSAppSyncClient.swift

```diff
@@ -918,13 +930,88 @@ public class AWSAppSyncClient {
         }
     }
     
+    /// Performs a sync operation where a base query is periodically called to fetch primary data from the server based on the syncConfiguration.
+    ///
+    /// - Parameters:
+    ///   - baseQuery: The base query to fetch which contains the primary data.
+    ///   - baseQueryResultHandler: Closure that is called when base query results are available or when an error occurs. Every time a sync operation is called, a fetch for the baseQuery from the cache will be done first before initiating any other operations.
+    ///   - deltaQuery: The delta query which fetches data starting from the `lastSync` time. Defaults to a no-op query.
+    ///   - deltaQueryResultHandler: Closure that is called when delta query executes. Defaults to a no-op handler.
+    ///   - callbackQueue: An optional queue on which sync callbacks will be invoked. Defaults to the main queue.
+    ///   - subscriptionWithSyncConfiguration: The sync configuration where the baseQuery sync interval can be specified. (Defaults to 24 hours.)
+    /// - Returns: An object that can be used to cancel the sync operation.
+    public func sync<BaseQuery: GraphQLQuery, DeltaQuery: GraphQLQuery>(baseQuery: BaseQuery,
+                                                                        baseQueryResultHandler: @escaping OperationResultHandler<BaseQuery>,
+                                                                        deltaQuery: DeltaQuery? = nil,
```

Comment:
marking `deltaQuery? = nil` will probably cause compilation error if un-specified.

Reply:
it should not, since default value is `nil`.

Reply:
The generic type would not be able to inferred causing the error I think.

Reply:
I'm not sure about this. It may be compiler bug if nil doesn't work. Did you check it?

Reply:
It definitely compiled fine for me locally. I'll try it in a few other contexts.

Reply:
It compiled OK for me, but in the interest of making the API very explicit about the supported use cases, I've made these required instead of optional. The three `sync` flavors are:
1. base only
1. base plus delta
1. base plus delta plus subscription

By not making the `delta` optional, it will make it clearer to callers that these are the three supported use cases (at the expense of a larger `sync` method family).